Add ZK opcodes definition (#619)
FuelLabs/fuel-specs: src/fuel-vm/instruction-set.md

- `$rB + $rC` overflows or `> VM_MAX_RAM`
- The memory range `MEM[$rA, 32]`  does not pass [ownership check](./index.md#ownership)

### `ECOP`: Elliptic curve point operation

|             |                                                     |
|-------------|-----------------------------------------------------|
| Description | This opcode regroups all arithmetic operations that can be perform on elliptic curve points. `$rB` defines the curve used. `$rC` defines the type of operation to perform. `$rD` points to the start of the bytes of the operation inputs in memory. `$rA` points to the start of the bytes of the result in memory.                                                   |
| Operation   | ```MEM[$rA, X] = ecop(MEM[$rD, Y]);```              |

> | Syntax      | `eadd $rA, $rB, $rC, $rD`                           |

Reviewer: should it be ecop here?

```suggestion
| Syntax      | `ecop $rA, $rB, $rC, $rD`                           |
```

| Encoding    | `0x00 rA rB rC rD`                                  |
| Notes       | For now, only `$rB` = 0 is accepted                 |

#### Curve ID `$rB` possible values

- `0`: `alt_bn128` elliptic curve.

#### Operation type `$rC` supported

- `0`: two points addition
- `1`: one point and one scalar multiplication

#### Encoding of points and results by curve ID and operation type

- 1P = one point = (X, Y) = ([32 bytes], [32 bytes])
- 1S = one scalar = X = [32 bytes]

| `$rB` Curve ID | `$rC` Operation type | `$rA` format         | `$rD` format               |
|----------------|----------------------|----------------------|----------------------------|

> |    `0`         | `0`                  | `MEM[$rA, 64]` `1P`  | `MEM[$rC, 128]` `1P1P`     |

Reviewer: ```suggestion
|    `0`         | `0`                  | `MEM[$rA, 64]` `1P`  | `MEM[$rC, 128]` `1P1S`     |
```

- `$rB + $rC` overflows or `> VM_MAX_RAM`
- The memory range `MEM[$rA, 32]`  does not pass [ownership check](./index.md#ownership)

### `ECOP`: Elliptic curve point operation

|             |                                                     |
|-------------|-----------------------------------------------------|

> | Description | This opcode regroups all arithmetic operations that can be perform on elliptic curve points. `$rB` defines the curve used. `$rC` defines the type of operation to perform. `$rD` points to the start of the bytes of the operation inputs in memory. `$rA` points to the start of the bytes of the result in memory.                                                   |

Reviewer: ```suggestion
| Description | This opcode regroups all arithmetic operations that can be performed on elliptic curve points. `$rB` defines the curve used. `$rC` defines the type of operation to perform. `$rD` points to the start of the bytes of the operation inputs in memory. `$rA` points to the start of the bytes of the result in memory.                                                   |
```

> | Description | This opcode regroups all arithmetic operations that can be performed on elliptic curve points. `$rB` defines the curve used. `$rC` defines the type of operation to perform. `$rD` points to the start of the bytes of the operation inputs in memory. `$rA` points to the start of the bytes of the result in memory.                                                   |

Reviewer 2: ```suggestion
| Description | Perform arithmetic operation `$rC` on points of the elliptic curve `$rB`. Arguments are read from memory at `$rD`, and is result written to the memory at `$rA`, as per the table below.                                                   |
```

Reviewer 2: Not sure if my wording is exact enough, but I felt like yours was unnecessarily verbose.

| Operation   | ```MEM[$rA, X] = ecop(MEM[$rD, Y]);```              |
| Syntax      | `ecop $rA, $rB, $rC, $rD`                           |
| Encoding    | `0x00 rA rB rC rD`                                  |
| Notes       | For now, only `$rB` = 0 is accepted                 |

#### Curve ID `$rB` possible values

- `0`: `alt_bn128` elliptic curve.

#### Operation type `$rC` supported

- `0`: two points addition
- `1`: one point and one scalar multiplication

#### Encoding of points and results by curve ID and operation type

- 1P = one point = (X, Y) = ([32 bytes], [32 bytes])
- 1S = one scalar = X = [32 bytes]

| `$rB` Curve ID | `$rC` Operation type | `$rA` format         | `$rD` format               |
|----------------|----------------------|----------------------|----------------------------|
|    `0`         | `0`                  | `MEM[$rA, 64]` `1P`  | `MEM[$rC, 128]` `1P1P`     |
|    `0`         | `1`                  | `MEM[$rA, 64]` `1P`  | `MEM[$rC, 96]` `1P1S`      |

### `EPAR`: Elliptic curve point pairing check

|             |                                                     |
|-------------|-----------------------------------------------------|

> | Description | Perform a specific pairing type within a specific curve both identified by `$rB`. `$rC` defines the number of batch of groups. `$rD` define where the bytes of the groups of points start. `$rA` contains either `0` or `1` as the result of the pairing. |

Reviewer 2: ```suggestion
| Description | Check if `$rC` groups of points at `$rD` all form valid pairings in (curve, pairing type) identified by `$rB`. Set `$rA` to the result of the pairing, either `0` or `1`. |
```
